datastream-server/src/main/java/com/linkedin/datastream/server/assignment/LoadBasedPartitionAssigner.java

```diff
@@ -170,4 +203,79 @@ int findTaskWithRoomForAPartition(List<String> tasks, Map<String, Set<String>> p
     }
     throw new DatastreamRuntimeException("No tasks found that can host an additional partition");
   }
+
+  void updateMetricsForDatastream(String datastream, PartitionAssignmentStats stats) {
+    if (!_partitionAssignmentStatsMap.containsKey(datastream)) {
+      registerLoadBasedPartitionAssignmentMetrics(datastream);
+    }
+    _partitionAssignmentStatsMap.put(datastream, stats);
+  }
+
+  private void registerLoadBasedPartitionAssignmentMetrics(String datastream) {
+    Supplier<Integer> minPartitionsAcrossTasksSupplier = () -> _partitionAssignmentStatsMap
+        .getOrDefault(datastream, PartitionAssignmentStats.DEFAULT).getMinPartitionsAcrossTasks();
+    DYNAMIC_METRICS_MANAGER.registerGauge(CLASS_NAME, datastream, MIN_PARTITIONS_ACROSS_TASKS,
+        minPartitionsAcrossTasksSupplier);
+
+    Supplier<Integer> maxPartitionsAcrossTasksSupplier = () -> _partitionAssignmentStatsMap
+        .getOrDefault(datastream, PartitionAssignmentStats.DEFAULT).getMaxPartitionsAcrossTasks();
+    DYNAMIC_METRICS_MANAGER.registerGauge(CLASS_NAME, datastream, MAX_PARTITIONS_ACROSS_TASKS,
+        maxPartitionsAcrossTasksSupplier);
+  }
+
+  /**
+   * {@inheritDoc}
+   */
+  @Override
+  public List<BrooklinMetricInfo> getMetricInfos() {
+    String prefix = CLASS_NAME + MetricsAware.KEY_REGEX;
+    return Arrays.asList(new BrooklinGaugeInfo(prefix + MIN_PARTITIONS_ACROSS_TASKS),
+        new BrooklinGaugeInfo(prefix + MAX_PARTITIONS_ACROSS_TASKS));
+  }
+
+  void cleanupMetrics() {
+    _partitionAssignmentStatsMap.keySet().forEach(this::unregisterMetricsForDatastream);
+    _partitionAssignmentStatsMap.clear();
+  }
+
+  void unregisterMetricsForDatastream(String datastream) {
+    DYNAMIC_METRICS_MANAGER.unregisterMetric(CLASS_NAME, datastream, MIN_PARTITIONS_ACROSS_TASKS);
+    DYNAMIC_METRICS_MANAGER.unregisterMetric(CLASS_NAME, datastream, MAX_PARTITIONS_ACROSS_TASKS);
+  }
+
+  /**
+   * Encapsulates assignment metrics for a single datastream group
+   */
+  private static class PartitionAssignmentStats {
+    private final int _minPartitionsAcrossTasks;
+    private final int _maxPartitionsAcrossTasks;
+
+    public static final PartitionAssignmentStats DEFAULT = new PartitionAssignmentStats(0, 0);
+
+    /**
+     * Creates an instance of {@link PartitionAssignmentStats}
```

Comment:
can we add a comment here to say that this class is for 'per-datastream metrics'

Reply:
I think the javadoc on the class already conveys this:
`Encapsulates assignment metrics for a single datastream group`